Presto properties documentation

anusudarsan opened the pull request:
supersedes https://github.com/prestodb/presto/pull/5753 . @kokosing

kokosing commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -410,6 +407,333 @@ Drop a schema::
+ * **Description:**
+
+  Limit on the nubmer of splits waiting to be served by a split source. After reaching
+  this limit, writers will stop writing new splits until some of hteme are used by workers.
```
writers? does it only related to `INSERT` queries? It is a bit confusing. I thought it is about readers which will not produce any new splits in there are many waiting to be processed, am I right?

anusudarsan replied:
Right. its not related to INSERTs (final writes) but to the un-scheduled splits waiting to be read.

kokosing commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -410,6 +407,333 @@ Drop a schema::
+
+  Limit on the nubmer of splits waiting to be served by a split source. After reaching
+  this limit, writers will stop writing new splits until some of hteme are used by workers.
+  Higher values will increase memory usage, but allow IO to be concentrated at one time,
```
Is IO concentration a good thing? Does it have any impact? Is it not about the data buffering? IMO as long buffer do not get empty during query processing then extending it should have not have any impact on query performance.

I think I would just omit this sentence with increasing the value.

kokosing commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -410,6 +407,333 @@ Drop a schema::
+  Maximum number of partitions per writer. A query will fail if it requires more partitions
+  per writer than allowed by this property. It can be helpful to have queries beyond the
+  expected maximum partitions to fail to help with error detection. Also it may allow to
+  preactivly avoid out of memory problem.
```
s/preactivly/proactively - typo

kokosing commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -410,6 +407,333 @@ Drop a schema::
+ * **Default value:** ``100``
+ * **Description:**
+
+  Maximum number of partitions per writer. A query will fail if it requires more partitions
```
isn't it about the thing that each writer has to have buffer per each partition and it takes memory so we need to have memory limit set? IMO description of this property does not match the real motivation behind it. Maybe we could ask someone from FB to explain its meaning, or we should dive into a code to find out by our self.

kokosing commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -410,6 +407,333 @@ Drop a schema::
+ * **Description:**
+
+  This property describes how many threads may be used to iterate through splits when loading
+  them to the worker nodes. A higher value may increase parallelism, but increased concurrency
```
I would drop: `A higher value may increase parallelism, but increased concurrency
  may cause too much time to be spent on context switching.`. It is obvious. 

Also here I would like to ask someone from FB about it, or dive into code.

kokosing commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -410,6 +407,333 @@ Drop a schema::
+ * **Description:**
+
+  The maximum size of splits created after the initial splits. The logic for initial splits is
+  described in ``hive.max-initial-splits``. A higher value will reduce parallelism. This may be
```
place `ref:` along with the name of related property

kokosing commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -410,6 +407,333 @@ Drop a schema::
+  sizes read from Hive. The first partition is always of size ``hive.metastore.partition-batch-size.min``
+  and each following partition is two times bigger than previous up to
+  ``hive.mestastore.partition-batch-size.max`` (the formula for partition size ``n`` is
+  min(``hive.metastore.partition-batch-size.max``, (``2``^``n``) * ``hive.metastore.partition-batch-size.min``)).
```
formula is not needed, the sentence is clear enough IMO

kokosing commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -410,6 +407,333 @@ Drop a schema::
+  If the queries in the system will differ significantly from each other in size, then this range should be
+  extended to better adjust to processing requirements. If the queries in the system will mostly be of the
+  same size, then setting both values to the same maximally tuned value may give a slight edge in
+  processing time.
```
I think it is also about responsiveness, if data to be read is small then it will be faster to load more smaller splits and execute the concurrently, but if query is processing big queries then it is better to have bigger splits (non iteractive query case).

anusudarsan replied:
ya but thats obvious right? Do we want to add it to the description?

kokosing replied:
ah... ok. Lets leave it as it is then.

kokosing commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -410,6 +407,333 @@ Drop a schema::
+ * **Default value:** ``false``
+ * **Description:**
+
+  Access ORC columns using names from the file. By default, Hive access columns in ORC files
```
s/access/accesses?

kokosing commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -410,6 +407,333 @@ Drop a schema::
+ * **Default value:** ``false``
+ * **Description:**
+
+  *Deprecated* Serves as default value for ``parquet_optimized_reader_enabled`` session property.
```
Why is it deprecated? Is going to be removed, is there any alternative? Description suggests is it more experimental than deprecated

anusudarsan replied:
not sure. looking at the code there is no alternative. I also think it was meant to be "experimental". @zhenxiao any idea?

zhenxiao replied:
parquet_optimized_reader_enabled is going to be turned on by default. Now due to a bug for complex types for Parquet, this property is disabled by default. After fixing that bug, which might cost some time, we will consider deprecate this property and turn it on by default.

kokosing commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -410,6 +407,333 @@ Drop a schema::
+ * **Description:**
+
+  Access ORC columns using names from the file. By default, Hive access columns in ORC files
+  using the order recoded in the Hive metastore. Setting this property allows to use columns
```
what does the `order` refer to? Is refers to order of how columns (column names) are stored in a file? Is there any impact on query performance? If so, when and why?

anusudarsan replied:
changed the description a bit..there is no perf impact for this change

kokosing commented on presto-docs/src/main/sphinx/admin/tuning.rst:
```diff
@@ -5,6 +5,28 @@ Tuning Presto
+  * :ref:`query.initial-hash-partitions<query-properties>`
+  * :ref:`task.concurrency<task-properties>`
+
+Those and other Presto properties are described in :doc:`properties article<properties>`.
```
s/article/section? It is not a blog or news service ;)

kokosing commented on presto-docs/src/main/sphinx/admin/tuning.rst:
```diff
@@ -5,6 +5,28 @@ Tuning Presto
+  * `query.max-memory = 200GB`
+  * `query.max-memory-per-node = 32GB`
+  * `query.initial-hash-partitions = 10`
+  * `task.concurrency = 8`
```
maybe some explanation how it was calculated would be beneficial to user as well. So she/he could calculate these numbers in her/his case.

kokosing commented on presto-docs/src/main/sphinx/admin/tuning.rst:
```diff
@@ -24,6 +24,13 @@ As an example, on a 11-node cluster dedicated to Presto (1 Coordinator + 10 Work
 
+If your workload consists of queries that have very quick processing time of the splits you might encounter situation when whole cluster is under utilized (very small usage of CPU, network and disks).
+For example this can happen when your queries are highly selective and most splits can be filtered out just by looking into theirs headers (like bloom filters or min/max statistics in ORC files) without reading the actual data.
+In such case you might want to increase values of the following properties:
```
why? what for?

anusudarsan replied:
sorry, didnt understand your question. You mean why this particular troubleshooting tip?

kokosing replied:
There is a case described above and there is a suggestion to user what should be tuned, but there is no explanation how this change is going influence the system. It would be nice to say to user how this will help.

kokosing commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -78,6 +78,14 @@ Spilling Properties
+
+ * **Type:** ``Boolean``
+ * **Default value:** ``true``
+ * **Description:** Allow reordering windows in order to put those with the same partitioning next to each other. This will sometimes decrease the number of repartitionings. This can also be specified on a per-query basis using the ``reorder_windows`` session property.
```
Either here or in other places is wrong formatting. Please take a look how David formats documentation. 

Here you have:
```
* **Description:** text....
```
while in other places you have:
```
* **Description:** 

text....
```

electrum replied:
Notes about changes:

* All text under the property header (including the bullet list) is indented 4 spaces
* No *"Description"* bullet
* The "type" name is lowercase. Use `data size` or `duration` rather than `string` where appropriate.
* Don't put a space when writing data size value. For example, write `32MB`, not `32 MB`
* Add an *"Allowed values"* bullet for `string` types that only allow certain values
* Add a *"Minimum value"* or *"Maximum value"* bullet when appropriate

kokosing commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -78,6 +78,14 @@ Spilling Properties
+
+ * **Type:** ``Boolean``
+ * **Default value:** ``true``
+ * **Description:** Allow reordering windows in order to put those with the same partitioning next to each other. This will sometimes decrease the number of repartitionings. This can also be specified on a per-query basis using the ``reorder_windows`` session property.
```
to me the word `sometimes` means it very rare case. I would prefer the word `may` as it means it is possible. A bit more positive. 

I would say something about possible benefits and costs.

electrum commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -515,3 +741,137 @@ The following properties allow tuning the :doc:`/functions/regexp`.
     correct algorithm from the beginning so as not to waste time and resources.
     The more rows you are processing, the larger this value should be.
+
+Session properties
```
This page is already quite long (and getting longer), so we should probably create a new `session-properties` page for these.

electrum commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -515,3 +741,137 @@ The following properties allow tuning the :doc:`/functions/regexp`.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+ * **Type:** ``String`` (``disabled``, ``columnar`` or ``columnar_dictionary``)
+ * **Default value:** ``optimizer.processing-optimization`` (``false``)
```
For all of these session properties that have config parameters as a default, let's write them like this:
```
* **Default value:** config ``optimizer.processing-optimization``
```
This makes it clear that it comes from a config property (especially important if we move session properties to a separate page). It also removes the duplicate default value, which is documented with the config property. I'm concerned that people will forget about this extra place when the config default is updated.

anusudarsan commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -76,6 +76,218 @@ Spilling Properties
+ * **Description:**
+
+  The path to the queue config file. Queues are used to manage the number of concurrent queries across the
+  system. More information on queues and how to configure them can be found in :doc:/admin/queue.
```
I think both usages are fine.

anusudarsan commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -410,6 +407,333 @@ Drop a schema::
+ * **Description:**
+
+  Maximum number of ranges/values allowed while reading hive data without compacting it.
+  A higher value will cause more data fragmentation but allow the use of the row skipping
```
reworded

anusudarsan commented:
@kokosing @electrum I think I addressed most of the comments. I squashed all the formatting changes to the original and kept only the actual content changes in the fixup commits.

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+
+.. _query-properties:
+
+Query execution properties
```
`Query Execution Properties` (seems like we use uppercase in other places)

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    * **Default value:** ``all-at-once``
+
+    Setting this value to ``phased`` will allow the query scheduler to split a single
+    query execution between different time slots. This will allow Presto to switch context
```
`switch context` -> what do you mean here? I found it confusing.

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    * **Default value:** ``100``
+
+    This value is used to determine how many nodes may share the same query when fixed
+    partitioning is chosen by Presto. Manipulating this value will affect the distribution
```
`Manipulating` -> `Changing` or `Updating`

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    of the cluster in a low traffic environment. An excessively high value will cause multiple
+    partitions of the same query to be assigned to a single node, or Presto may ignore
+    the setting if ``node-scheduler.multiple-tasks-per-node-enabled`` is set to false -
+    the value is internally capped at the number of available worker nodes in such scenario.
```
`in such scenario` -> `in that case` or `in such a scenario` ?

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    * **Default value:** ``false``
+
+    This property controls whether a query killer should be triggered when a cluster
+    is running out of memory. The killer will drop the largest queries first so enabling
```
`drop` -> `kill`

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+
+    This property controls whether a query killer should be triggered when a cluster
+    is running out of memory. The killer will drop the largest queries first so enabling
+    this option may cause problems with executing large queries in a highly loaded cluster,
```
`with` -> `when`

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    This property controls whether a query killer should be triggered when a cluster
+    is running out of memory. The killer will drop the largest queries first so enabling
+    this option may cause problems with executing large queries in a highly loaded cluster,
+    but should increase stability of smaller queries.
```
`it should increase the stability`

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+
+    Size of the thread pool used for garbage collecting after queries. Threads from this
+    pool are used to free resources from canceled queries, as well as enforce memory limits,
+    queries timeouts etc. More threads will allow for more efficient memory management,
```
`queries timeouts etc.` -> `query timeouts, etc.`

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    Size of the thread pool used for garbage collecting after queries. Threads from this
+    pool are used to free resources from canceled queries, as well as enforce memory limits,
+    queries timeouts etc. More threads will allow for more efficient memory management,
+    and so may help avoid out of memory exceptions in some scenarios. However, having more
```
`scenarios` -> `cases`.

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    pool are used to free resources from canceled queries, as well as enforce memory limits,
+    queries timeouts etc. More threads will allow for more efficient memory management,
+    and so may help avoid out of memory exceptions in some scenarios. However, having more
+    threads may also increase CPU usage for garbage collecting and will have an additional
```
`CPU usage for garbage collecting` -> this is confusing and made me think about the JVM garbage collection.

I feel like we shouldn't use the phrase "garbage collection" here as it may be confusing for some people.

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+
+    This property describes the minimum time after which the query metadata may be removed
+    from the server. If the value is too low, the client may not be able to receive information
+    about query completion. The value describes minimum time, but if there is space available
```
`The value of this property sets the minimum time` ?

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    * **Type:** ``data size``
+    * **Default value:** ``20GB``
+
+    Serves as the default value for the ``query_max_memory`` session property. This property also describes
```
`describes` -> `configures` / `sets`

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    * **Default value:** ``20GB``
+
+    Serves as the default value for the ``query_max_memory`` session property. This property also describes
+    the strict limit of total memory that may be used to process a single query. A query is dropped if the
```
`dropped` -> `killed`

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    * **Minimum Value:** ``1``
+    * **Default value:** ``1000``
+
+    This value describes the maximum size of the thread pool used to handle responses to HTTP requests for
```
`handle responses to HTTP requests` -> `run callbacks generated while processing HTTP requests`?

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+
+    This value describes the maximum size of the thread pool used to handle responses to HTTP requests for
+    each task. Increasing this value will cause more resources to be used for handling HTTP communication
+    itself, but may also improve response time when Presto is distributed across many hosts or there are
```
`response time` -> `the response time`

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    This value describes the maximum size of the thread pool used to handle responses to HTTP requests for
+    each task. Increasing this value will cause more resources to be used for handling HTTP communication
+    itself, but may also improve response time when Presto is distributed across many hosts or there are
+    a lot of small queries being run.
```
>> a lot of small queries being run.

It would be nice to rephrase this. First, `small queries` is a little bit vague here, I think you mean in terms of the number of tasks it creates, so maybe we should rephrase this a bit to make it explicit. 

Also the end of sentence `being run` sounds weird.

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    * **Minimum Value:** ``1s``
+    * **Default value:** ``2m``
+
+    The minimal time that HTTP worker must be unavailable before the coordinator assumes the worker crashed.
```
`minimal` -> `minimum`

nezihyigitbasi replied:
what is `HTTP worker`?

nezihyigitbasi replied:
`assumes the worker crashed` -> `assumes that the worker is crashed`

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    * **Default value:** ``2m``
+
+    The minimal time that HTTP worker must be unavailable before the coordinator assumes the worker crashed.
+    A higher value may be recommended in unstable connection conditions. In order to consider a node dead,
```
what does `unstable connection conditions.` mean?

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+
+    The minimal time that HTTP worker must be unavailable before the coordinator assumes the worker crashed.
+    A higher value may be recommended in unstable connection conditions. In order to consider a node dead,
+    the defined time must pass between two failed attempts of HTTP communication, with no successful communication
```
`defined` -> `configured`

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    * **Default value:** ``1000``
+
+    The number of splits that will be attempted to fetch in a single stage. Higher value may
+    be used if system works in reliable environment and the responsiveness is less important then average answer
```
`then the average answer time` -> `than the average response time`

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+
+    The number of splits that will be attempted to fetch in a single stage. Higher value may
+    be used if system works in reliable environment and the responsiveness is less important then average answer
+    time, it will require more memory reserve though. Decreasing this value may have a positive effect if
```
`more memory reserve` -> `more memory`

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/properties.rst:
```diff
@@ -58,6 +60,251 @@ General Properties
+    The number of splits that will be attempted to fetch in a single stage. Higher value may
+    be used if system works in reliable environment and the responsiveness is less important then average answer
+    time, it will require more memory reserve though. Decreasing this value may have a positive effect if
+    there are lots of nodes in system and calculations are relatively heavy for each of splits. If the number is too
```
`system` -> `the cluster`

nezihyigitbasi replied:
`calculations are relatively heavy for each of splits.` -> `processing of each split is relatively expensive.`?

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/session-properties.rst:
```diff
@@ -0,0 +1,111 @@
+
+    Use resources that are not guaranteed to be available to a query. This property allows you to exceed
+    the limits of memory available per query and session. It may allow resources to be used more efficiently,
+    but may also cause non-deterministic query drops due to insufficient memory on machine. It can be
```
`drops` -> `failures`

nezihyigitbasi commented on presto-docs/src/main/sphinx/admin/session-properties.rst:
```diff
@@ -0,0 +1,111 @@
+    Use resources that are not guaranteed to be available to a query. This property allows you to exceed
+    the limits of memory available per query and session. It may allow resources to be used more efficiently,
+    but may also cause non-deterministic query drops due to insufficient memory on machine. It can be
+    particularly useful for performing more demanding queries.
```
`demanding` -> `resource intensive`?

anusudarsan commented:
@nezihyigitbasi thanks for the comments. Addressed them and rebased.

anusudarsan commented:
@electrum ping :)

electrum commented:
@raghavsethi is going to review this

electrum commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -108,6 +108,21 @@ The properties that apply to Hive connector security are listed in the
+directory is specified by the configuration variable ``hive.metastore.warehouse.dir``
+in ``hive-site.xml``, and the default value is ``/user/hive/warehouse``. If that
+is not the case, either add the following to ``jvm.config`` on all of the nodes:
+``-DHADOOP_USER_NAME=USER``, where ``USER`` is an operating system user that has proper
```
I understand documenting things as they are, but is there a way to make this a normal Hive config property?

anusudarsan replied:
@electrum yes. that is fine. I will drop that change from this PR, and come back with a code fix.

kokosing commented:
@raghavsethi Can you please take a look at this? We are going to get this through the pull request review process for more than a year, it would be lovely to get this finally merged.

kokosing commented:
@mattsfuller Can you please review this and check grammar and if description is clear enough.

raghavsethi commented:
This PR has too many examples of grammatical errors and confusing explanations to be accepted. In addition, I think we can omit documenting experimental or deprecated properties. While we're looking at these it might be wroth examining which of these properties are actually useful in production and whether we can actually remove some of them.

raghavsethi commented:
I talked to @kokosing offline and the conclusion was that I will revisit this PR after it has been updated.

raghavsethi commented:
Is this still being worked on?

kokosing commented:
I do not think so... other things got higher priority.

anusudarsan commented:
@raghavsethi  The PR has been updated with just the query config properties (cleaned up/removed deprecated properties, added configs implemented recently etc). Please take a look. 

Once we merge this, I will clean up the hive connector properties and open a new PR.
